File: backend/app/services/markdown_sections.py
import re

HEADING_RE = re.compile(r"^(#{1,6})\s+(.+?)\s*$", re.MULTILINE)


def normalize_heading_title(title: str) -> str:
    title = re.sub(r"^#+\s*", "", title)
    title = re.sub(r"^第[一二三四五六七八九十百千零〇\d]+[章节部分篇]\s*", "", title)
    title = re.sub(r"^(?:\d+(?:[.\-．]\d+)*)[、.．]?\s*", "", title)
    title = re.sub(r"^[（(][\d一二三四五六七八九十]+[）)]\s*", "", title)
    title = re.sub(r"[\s：:]+", "", title)
    return title.lower()


def titles_match(left: str, right: str) -> bool:
    a = normalize_heading_title(left)
    b = normalize_heading_title(right)
    return bool(a and b and a == b)


def find_markdown_sections(markdown: str) -> list[dict]:
    matches = [
        {"level": len(match.group(1)), "title": match.group(2).strip(), "start": match.start()}
        for match in HEADING_RE.finditer(markdown)
    ]
    sections = []
    for index, item in enumerate(matches):
        next_section = next(
            (other for other in matches[index + 1 :] if other["level"] <= item["level"]),
            None,
        )
        sections.append({**item, "end": next_section["start"] if next_section else len(markdown)})
    return sections


def extract_markdown_section(markdown: str, title: str, level: int | None = None) -> str | None:
    matches = [
        section
        for section in find_markdown_sections(markdown)
        if titles_match(section["title"], title)
    ]
    if not matches:
        return None
    section = next((item for item in matches if level is None or item["level"] == level), matches[0])
    return markdown[section["start"] : section["end"]].strip()

File: frontend/src/utils/markdownSections.ts
export function normalizeHeadingTitle(title: string): string {
  return title
    .replace(/^#+\s*/, "")
    .replace(/^第[一二三四五六七八九十百千零〇\d]+[章节部分篇]\s*/, "")
    .replace(/^(?:\d+(?:[.\-．]\d+)*)[、.．]?\s*/, "")
    .replace(/^[（(][\d一二三四五六七八九十]+[）)]\s*/, "")
    .replace(/[\s：:]+/g, "")
    .toLowerCase()
}

export function titlesMatch(left: string, right: string): boolean {
  const a = normalizeHeadingTitle(left)
  const b = normalizeHeadingTitle(right)
  return Boolean(a && b && a === b)
}

export interface MarkdownSection {
  level: number
  title: string
  start: number
  end: number
}

export function findMarkdownSections(markdown: string): MarkdownSection[] {
  const heading = /^(#{1,6})\s+(.+?)\s*$/gm
  const matches: { level: number; title: string; start: number }[] = []
  let match = heading.exec(markdown)
  while (match) {
    matches.push({
      level: match[1].length,
      title: match[2].trim(),
      start: match.index,
    })
    match = heading.exec(markdown)
  }
  return matches.map((item, index) => {
    const next = matches.slice(index + 1).find(other => other.level <= item.level)
    return {
      level: item.level,
      title: item.title,
      start: item.start,
      end: next ? next.start : markdown.length,
    }
  })
}

function pickSection(
  markdown: string,
  title: string,
  level?: number,
): MarkdownSection | null {
  const matches = findMarkdownSections(markdown).filter(section => titlesMatch(section.title, title))
  if (!matches.length) return null
  if (level == null) return matches[0]
  return matches.find(section => section.level === level) ?? matches[0]
}

export function extractMarkdownSection(
  markdown: string,
  title: string,
  level?: number,
): string | null {
  const section = pickSection(markdown, title, level)
  return section ? markdown.slice(section.start, section.end).trim() : null
}

export function replaceMarkdownSection(
  markdown: string,
  title: string,
  newSection: string,
  level?: number,
): string | null {
  const section = pickSection(markdown, title, level)
  if (!section) return null
  const replacement = newSection.trim()
  const before = markdown.slice(0, section.start)
  const after = markdown.slice(section.end)
  const prefix = before.length > 0 && !before.endsWith("\n") ? "\n" : ""
  const suffix = after.length > 0 && !replacement.endsWith("\n") ? "\n" : ""
  return `${before}${prefix}${replacement}${suffix}${after}`
}
